src/cmt_decode_statsd.c

```diff
@@ -0,0 +1,574 @@
+#include <cmetrics/cmt_histogram.h>
+#include <cmetrics/cmt_gauge.h>
+#include <cmetrics/cmt_untyped.h>
+#include <cmetrics/cmt_decode_statsd.h>
+#include <cmetrics/cmt_compat.h>
+
+static struct cmt_map_label *create_map_label(char *caption, size_t length)
+{
+    struct cmt_map_label *map_label;
+
+    map_label = calloc(1, sizeof(struct cmt_map_label));
+    if (!map_label) {
+        return NULL;
+    }
+
+    if (map_label != NULL) {
+        if (caption != NULL) {
+            if (length == 0) {
+                length = strlen(caption);
+            }
+
+            map_label->name = cfl_sds_create_len(caption, length);
+
+            if (map_label->name == NULL) {
+                cmt_errno();
+
+                free(map_label);
+
+                map_label = NULL;
+            }
+        }
+    }
+
+    return map_label;
+}
+
+static int append_new_map_label_key(struct cmt_map *map, char *name)
+{
+    struct cmt_map_label *label;
+
+    label = create_map_label(name, 0);
+
+    if (label == NULL) {
+        return CMT_DECODE_STATSD_ALLOCATION_ERROR;
+    }
+
+    cfl_list_add(&label->_head, &map->label_keys);
+    map->label_count++;
+
+    return CMT_DECODE_STATSD_SUCCESS;
+}
+
+static int append_new_metric_label_value(struct cmt_metric *metric, char *name, size_t length)
+{
+    struct cmt_map_label *label;
+
+    label = create_map_label(name, length);
+
+    if (label == NULL) {
+        return CMT_DECODE_STATSD_ALLOCATION_ERROR;
+    }
+
+    cfl_list_add(&label->_head, &metric->labels);
+
+    return CMT_DECODE_STATSD_SUCCESS;
+}
+
+static int is_incremental(char *str)
+{
+    return (*str == '+' || *str == '-');
+}
+
+static int decode_labels(struct cmt *cmt,
+                         struct cmt_map *map,
+                         struct cmt_metric *metric,
+                         char *labels, int incremental)
+{
+    void                 **value_index_list;
+    size_t                 map_label_index;
+    size_t                 map_label_count;
+    struct cfl_list       *label_iterator;
+    struct cmt_map_label  *current_label;
+    size_t                 label_index;
+    int                    label_found;
+    char                  *label_kv, *colon;
+    cfl_sds_t              label_k, label_v;
+    int                    result;
+    char                  *store;
+
+    result = CMT_DECODE_STATSD_SUCCESS;
+
+    value_index_list = calloc(128, sizeof(void *));
+
+    if (value_index_list == NULL) {
+        return CMT_DECODE_STATSD_ALLOCATION_ERROR;
+    }
+
+    label_found = CMT_FALSE;
+    label_index = 0;
+
+    if (incremental) {
+        label_k = cfl_sds_create("incremental");
+        result = append_new_map_label_key(map, label_k);
+        cfl_sds_destroy(label_k);
+
+        if (result == CMT_DECODE_STATSD_SUCCESS) {
+            value_index_list[label_index] = (void *) cfl_sds_create("true");
+        }
+    }
+
+    if (labels != NULL) {
+        label_kv = cmt_platform_strtok_r(labels, ",", &store);
```

Comment:
no strtok pls

Reply:
How could we handle on this? Should we use strspn or other string manipulating functions?